src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/ProgramDescriptor.java

```diff
@@ -46,30 +48,50 @@
  * risk.  This code and its internal interfaces are subject to change
  * or deletion without notice.</strong></p>
  */
-public record ProgramDescriptor(ProgramFileObject fileObject, Optional<String> packageName, Path sourceRootPath) {
+public record ProgramDescriptor(
+        ProgramFileObject fileObject,
+        Optional<String> packageName,
+        List<String> declaredTypeNames,
+        Path sourceRootPath) {
     static ProgramDescriptor of(ProgramFileObject fileObject) throws Fault {
         var file = fileObject.getFile();
+        var packageName = ""; // empty string will be converted into an empty optional
+        var declaredTypeNames = new ArrayList<String>();
         try {
             var compiler = JavacTool.create();
             var standardFileManager = compiler.getStandardFileManager(null, null, null);
             var units = List.of(fileObject);
             var task = compiler.getTask(null, standardFileManager, diagnostic -> {}, null, null, units);
             for (var tree : task.parse()) {
+                for (var decl : tree.getTypeDecls()) {
```

Comment:
Not terribly important, but you could first compute the package name, and then compute the declared types. That would permit creating the fully qualified `declaredTypeNames` in one step, rather than putting the simple names into a List and then going through them again to qualify them.

Reply:
Refactored as suggested.